File: cathode/src/main/java/net/simonvt/cathode/ui/person/PersonFragment.java
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.ui.person;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import java.util.List;
import javax.inject.Inject;
import net.simonvt.cathode.Injector;
import net.simonvt.cathode.R;
import net.simonvt.cathode.api.enumeration.Department;
import net.simonvt.cathode.api.enumeration.ItemType;
import net.simonvt.cathode.api.util.TraktUtils;
import net.simonvt.cathode.jobqueue.Job;
import net.simonvt.cathode.scheduler.MovieTaskScheduler;
import net.simonvt.cathode.scheduler.PersonTaskScheduler;
import net.simonvt.cathode.scheduler.ShowTaskScheduler;
import net.simonvt.cathode.settings.TraktTimestamps;
import net.simonvt.cathode.ui.LibraryType;
import net.simonvt.cathode.ui.NavigationListener;
import net.simonvt.cathode.ui.fragment.RefreshableAppBarFragment;
import net.simonvt.cathode.common.util.Ids;
import net.simonvt.cathode.common.util.Intents;
import net.simonvt.cathode.common.util.guava.Preconditions;
import net.simonvt.cathode.widget.RemoteImageView;

public class PersonFragment extends RefreshableAppBarFragment {

  private static final String TAG = "net.simonvt.cathode.ui.person.PersonFragment";

  private static final String ARG_PERSON_ID =
      "net.simonvt.cathode.ui.person.PersonFragment.personId";

  private static final int LOADER_PERSON = 1;

  @Inject PersonTaskScheduler personScheduler;
  @Inject ShowTaskScheduler showScheduler;
  @Inject MovieTaskScheduler movieScheduler;

  private long personId;

  private Person person;

  private int itemCount;

  @BindView(R.id.headshot) RemoteImageView headshot;
  @BindView(R.id.bornTitle) View bornTitle;
  @BindView(R.id.birthday) TextView birthday;
  @BindView(R.id.birthplace) TextView birthplace;
  @BindView(R.id.deathTitle) View deathTitle;
  @BindView(R.id.death) TextView death;
  @BindView(R.id.biography) TextView biography;

  @BindView(R.id.cast_header) LinearLayout castHeader;
  @BindView(R.id.cast_items) LinearLayout castItems;

  @BindView(R.id.production_header) LinearLayout productionHeader;
  @BindView(R.id.production_items) LinearLayout productionItems;

  @BindView(R.id.art_header) LinearLayout artHeader;
  @BindView(R.id.art_items) LinearLayout artItems;

  @BindView(R.id.crew_header) LinearLayout crewHeader;
  @BindView(R.id.crew_items) LinearLayout crewItems;

  @BindView(R.id.costume_makeup_header) LinearLayout costumeMakeupHeader;
  @BindView(R.id.costume_makeup_items) LinearLayout costumeMakeupItems;

  @BindView(R.id.directing_header) LinearLayout directingHeader;
  @BindView(R.id.directing_items) LinearLayout directingItems;

  @BindView(R.id.writing_header) LinearLayout writingHeader;
  @BindView(R.id.writing_items) LinearLayout writingItems;

  @BindView(R.id.sound_header) LinearLayout soundHeader;
  @BindView(R.id.sound_items) LinearLayout soundItems;

  @BindView(R.id.camera_header) LinearLayout cameraHeader;
  @BindView(R.id.camera_items) LinearLayout cameraItems;

  @BindView(R.id.viewOnTrakt) View viewOnTrakt;

  private NavigationListener navigationListener;

  public static String getTag(long personId) {
    return TAG + "/" + personId + "/" + Ids.newId();
  }

  public static Bundle getArgs(long personId) {
    Preconditions.checkArgument(personId >= 0, "personId must be >= 0");

    Bundle args = new Bundle();
    args.putLong(ARG_PERSON_ID, personId);
    return args;
  }

  @Override public void onAttach(Activity activity) {
    super.onAttach(activity);
    navigationListener = (NavigationListener) activity;
  }

  @Override public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Injector.obtain().inject(this);

    Bundle args = getArguments();
    personId = args.getLong(ARG_PERSON_ID);

    itemCount = getResources().getInteger(R.integer.personCreditColumns);

    getLoaderManager().initLoader(LOADER_PERSON, null, personLoader);
  }

  @Override public void onRefresh() {
    personScheduler.sync(personId, onDoneListener);
  }

  private Job.OnDoneListener onDoneListener = new Job.OnDoneListener() {
    @Override public void onDone(Job job) {
      setRefreshing(false);
    }
  };

  @Override
  protected View createView(LayoutInflater inflater, ViewGroup container, Bundle inState) {
    return inflater.inflate(R.layout.fragment_person, container, false);
  }

  @Override public void onViewCreated(View view, Bundle inState) {
    super.onViewCreated(view, inState);
    updateView(person);
  }

  @OnClick(R.id.cast_header) void onDisplayCastCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.CAST);
  }

  @OnClick(R.id.production_header) void onDisplayProductionCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.PRODUCTION);
  }

  @OnClick(R.id.art_header) void onDisplayArtCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.ART);
  }

  @OnClick(R.id.crew_header) void onDisplayCrewCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.CREW);
  }

  @OnClick(R.id.costume_makeup_header) void onDisplayCostumeMakeUpCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.COSTUME_AND_MAKEUP);
  }

  @OnClick(R.id.directing_header) void onDisplayDirectingCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.DIRECTING);
  }

  @OnClick(R.id.writing_header) void onDisplayWritingCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.WRITING);
  }

  @OnClick(R.id.sound_header) void onDisplaySoundCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.SOUND);
  }

  @OnClick(R.id.camera_header) void onDisplayCameraCredits() {
    navigationListener.onDisplayPersonCredit(personId, Department.CAMERA);
  }

  private void updateView(final Person person) {
    this.person = person;
    if (person != null && getView() != null) {
      setTitle(person.getName());
      setBackdrop(person.getScreenshot());
      headshot.setImage(person.getHeadshot());

      if (!TextUtils.isEmpty(person.getBirthday())) {
        bornTitle.setVisibility(View.VISIBLE);
        birthday.setVisibility(View.VISIBLE);
        birthplace.setVisibility(View.VISIBLE);

        birthday.setText(person.getBirthday());
        birthplace.setText(person.getBirthplace());
      } else {
        bornTitle.setVisibility(View.GONE);
        birthday.setVisibility(View.GONE);
        birthplace.setVisibility(View.GONE);
      }

      if (!TextUtils.isEmpty(person.getDeath())) {
        deathTitle.setVisibility(View.VISIBLE);
        death.setVisibility(View.VISIBLE);
        death.setText(person.getDeath());
      } else {
        deathTitle.setVisibility(View.GONE);
        death.setVisibility(View.GONE);
      }

      biography.setText(person.getBiography());

      updateItems(castHeader, castItems, person.getCredits().getCast());
      updateItems(productionHeader, productionItems, person.getCredits().getProduction());
      updateItems(artHeader, artItems, person.getCredits().getArt());
      updateItems(crewHeader, crewItems, person.getCredits().getCrew());
      updateItems(costumeMakeupHeader, costumeMakeupItems,
          person.getCredits().getCostumeAndMakeUp());
      updateItems(directingHeader, directingItems, person.getCredits().getDirecting());
      updateItems(writingHeader, writingItems, person.getCredits().getWriting());
      updateItems(soundHeader, soundItems, person.getCredits().getSound());
      updateItems(cameraHeader, cameraItems, person.getCredits().getCamera());

      viewOnTrakt.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          Intents.openUrl(getContext(), TraktUtils.getTraktPersonUrl(person.getTraktId()));
        }
      });

      if (TraktTimestamps.shouldSyncPerson(person.getLastSync())) {
        personScheduler.sync(personId);
      }
    }
  }

  private void updateItems(View header, ViewGroup items, List<PersonCredit> credits) {
    items.removeAllViews();

    final int size = credits != null ? credits.size() : 0;
    if (size > 0) {
      header.setVisibility(View.VISIBLE);
      items.setVisibility(View.VISIBLE);

      for (int i = 0; i < size && i < itemCount; i++) {
        final PersonCredit credit = credits.get(i);

        View view =
            LayoutInflater.from(getContext()).inflate(R.layout.person_item_credit, items, false);

        RemoteImageView poster = ButterKnife.findById(view, R.id.poster);
        TextView title = ButterKnife.findById(view, R.id.title);
        TextView job = ButterKnife.findById(view, R.id.job);

        poster.setImage(credit.getPoster());
        title.setText(credit.getTitle());

        if (credit.getJob() != null) {
          job.setText(credit.getJob());
        } else {
          job.setText(credit.getCharacter());
        }

        view.setOnClickListener(new View.OnClickListener() {
          @Override public void onClick(View v) {
            if (credit.getItemType() == ItemType.SHOW) {
              navigationListener.onDisplayShow(credit.getItemId(), credit.getTitle(),
                  credit.getOverview(), LibraryType.WATCHED);
            } else {
              navigationListener.onDisplayMovie(credit.getItemId(), credit.getTitle(),
                  credit.getOverview());
            }
          }
        });

        items.addView(view);
      }
    } else {
      header.setVisibility(View.GONE);
      items.setVisibility(View.GONE);
    }
  }

  private LoaderManager.LoaderCallbacks<Person> personLoader =
      new LoaderManager.LoaderCallbacks<Person>() {
        @Override public Loader<Person> onCreateLoader(int id, Bundle args) {
          return new PersonLoader(getContext(), personId);
        }

        @Override public void onLoadFinished(Loader<Person> loader, Person data) {
          updateView(data);
        }

        @Override public void onLoaderReset(Loader<Person> loader) {
        }
      };
}

File: cathode/src/main/java/net/simonvt/cathode/ui/person/PersonFragment.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.person

import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.vectordrawable.graphics.drawable.VectorDrawableCompat
import net.simonvt.cathode.R
import net.simonvt.cathode.api.enumeration.Department
import net.simonvt.cathode.api.enumeration.ItemType
import net.simonvt.cathode.api.util.TraktUtils
import net.simonvt.cathode.common.ui.fragment.RefreshableAppBarFragment
import net.simonvt.cathode.common.util.Ids
import net.simonvt.cathode.common.util.Intents
import net.simonvt.cathode.common.util.guava.Preconditions
import net.simonvt.cathode.databinding.FragmentPersonBinding
import net.simonvt.cathode.databinding.PersonItemCreditBinding
import net.simonvt.cathode.ui.CathodeViewModelFactory
import net.simonvt.cathode.ui.LibraryType
import net.simonvt.cathode.ui.NavigationListener
import javax.inject.Inject

class PersonFragment @Inject constructor(
  private val viewModelFactory: CathodeViewModelFactory
) : RefreshableAppBarFragment() {

  private var personId: Long = -1L

  private val viewModel: PersonViewModel by viewModels { viewModelFactory }

  private var person: Person? = null

  private var itemCount: Int = 0

  private var _binding: FragmentPersonBinding? = null
  private val binding get() = _binding!!

  lateinit var navigationListener: NavigationListener

  override fun onAttach(context: Context) {
    super.onAttach(context)
    navigationListener = requireActivity() as NavigationListener
  }

  override fun onCreate(inState: Bundle?) {
    super.onCreate(inState)
    personId = requireArguments().getLong(ARG_PERSON_ID)

    itemCount = resources.getInteger(R.integer.personCreditColumns)

    viewModel.setPersonId(personId)
    viewModel.loading.observe(this, Observer { loading -> setRefreshing(loading) })
    viewModel.person.observe(this, Observer { person -> updateView(person) })
  }

  override fun onRefresh() {
    viewModel.refresh()
  }

  override fun createView(inflater: LayoutInflater, container: ViewGroup?, inState: Bundle?): View {
    _binding = FragmentPersonBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, inState: Bundle?) {
    super.onViewCreated(view, inState)
    val linkDrawable = VectorDrawableCompat.create(resources, R.drawable.ic_link_black_24dp, null)
    binding.viewOnTrakt.setCompoundDrawablesWithIntrinsicBounds(linkDrawable, null, null, null)
    binding.castHeader.setOnClickListener(castHeaderClickListener)
    binding.productionHeader.setOnClickListener(productionHeaderClickListener)
    binding.artHeader.setOnClickListener(artHeaderClickListener)
    binding.crewHeader.setOnClickListener(crewHeaderClickListener)
    binding.costumeMakeupHeader.setOnClickListener(costumeMakeupHeaderClickListener)
    binding.directingHeader.setOnClickListener(directingHeaderClickListener)
    binding.writingHeader.setOnClickListener(writingHeaderClickListener)
    binding.soundHeader.setOnClickListener(soundHeaderClickListener)
    binding.cameraHeader.setOnClickListener(cameraHeaderClickListener)

    updateView(person)
  }

  override fun onDestroyView() {
    _binding = null
    super.onDestroyView()
  }

  private val castHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.CAST)
  }

  private val productionHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.PRODUCTION)
  }

  private val artHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.ART)
  }

  private val crewHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.CREW)
  }

  private val costumeMakeupHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.COSTUME_AND_MAKEUP)
  }

  private val directingHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.DIRECTING)
  }

  private val writingHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.WRITING)
  }

  private val soundHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.SOUND)
  }

  private val cameraHeaderClickListener = View.OnClickListener {
    navigationListener.onDisplayPersonCredit(personId, Department.CAMERA)
  }

  private fun updateView(person: Person?) {
    this.person = person
    if (person != null && view != null) {
      setTitle(person.name)
      setBackdrop(person.screenshot)
      binding.headshot.setImage(person.headshot)

      if (!TextUtils.isEmpty(person.birthday)) {
        binding.bornTitle.visibility = View.VISIBLE
        binding.birthday.visibility = View.VISIBLE
        binding.birthplace.visibility = View.VISIBLE

        binding.birthday.text = person.birthday
        binding.birthplace.text = person.birthplace
      } else {
        binding.bornTitle.visibility = View.GONE
        binding.birthday.visibility = View.GONE
        binding.birthplace.visibility = View.GONE
      }

      if (!TextUtils.isEmpty(person.death)) {
        binding.deathTitle.visibility = View.VISIBLE
        binding.death.visibility = View.VISIBLE
        binding.death.text = person.death
      } else {
        binding.deathTitle.visibility = View.GONE
        binding.death.visibility = View.GONE
      }

      binding.biography.text = person.biography

      updateItems(binding.castHeader, binding.castItems, person.credits.cast)
      updateItems(binding.productionHeader, binding.productionItems, person.credits.production)
      updateItems(binding.artHeader, binding.artItems, person.credits.art)
      updateItems(binding.crewHeader, binding.crewItems, person.credits.crew)
      updateItems(
        binding.costumeMakeupHeader,
        binding.costumeMakeupItems,
        person.credits.costumeAndMakeUp
      )
      updateItems(binding.directingHeader, binding.directingItems, person.credits.directing)
      updateItems(binding.writingHeader, binding.writingItems, person.credits.writing)
      updateItems(binding.soundHeader, binding.soundItems, person.credits.sound)
      updateItems(binding.cameraHeader, binding.cameraItems, person.credits.camera)

      binding.viewOnTrakt.setOnClickListener {
        Intents.openUrl(
          requireContext(),
          TraktUtils.getTraktPersonUrl(person.traktId)
        )
      }
    }
  }

  private fun updateItems(header: View?, items: ViewGroup, credits: List<PersonCredit>?) {
    items.removeAllViews()

    val size = credits?.size ?: 0
    if (size > 0) {
      header!!.visibility = View.VISIBLE
      items.visibility = View.VISIBLE

      var i = 0
      while (i < size && i < itemCount) {
        val credit = credits!![i]

        val itemBinding =
          PersonItemCreditBinding.inflate(LayoutInflater.from(requireContext()), items, false)

        itemBinding.poster.setImage(credit.getPoster())
        itemBinding.title.text = credit.getTitle()

        if (credit.getJob() != null) {
          itemBinding.job.text = credit.getJob()
        } else {
          itemBinding.job.text = credit.getCharacter()
        }

        itemBinding.root.setOnClickListener {
          if (credit.getItemType() === ItemType.SHOW) {
            navigationListener.onDisplayShow(
              credit.getItemId(),
              credit.getTitle(),
              credit.getOverview(),
              LibraryType.WATCHED
            )
          } else {
            navigationListener.onDisplayMovie(
              credit.getItemId(),
              credit.getTitle(),
              credit.getOverview()
            )
          }
        }

        items.addView(itemBinding.root)
        i++
      }
    } else {
      header!!.visibility = View.GONE
      items.visibility = View.GONE
    }
  }

  companion object {

    private const val TAG = "net.simonvt.cathode.ui.person.PersonFragment"

    private const val ARG_PERSON_ID = "net.simonvt.cathode.ui.person.PersonFragment.personId"

    @JvmStatic
    fun getTag(personId: Long): String {
      return TAG + "/" + personId + "/" + Ids.newId()
    }

    @JvmStatic
    fun getArgs(personId: Long): Bundle {
      Preconditions.checkArgument(personId >= 0, "personId must be >= 0")

      val args = Bundle()
      args.putLong(ARG_PERSON_ID, personId)
      return args
    }
  }
}
